File: cyrio-java/src/main/java/com/cyrio/core/model/Song.java
package com.cyrio.core.model;

/**
 * 歌曲信息（用户面向数据模型）
 *
 * <p>由 Rust cyrio-core 通过 JNI 返回的 JSON 数据解析而来。
 * {@code bitRate} 已在 Rust 端从设备内部的 {@code kbps << 7} 格式转换为 kbps。
 *
 * <p>仅保留用户关心的元数据，不含设备协议细节。
 */
public class Song {

    /** 文件编号（设备内部的唯一编号，用于下载/删除/加入歌单） */
    public int fileNo;

    /** 文件大小（字节） */
    public long size;

    /** 时长（秒） */
    public int time;

    /** 比特率（kbps） */
    public int bitRate;

    /** 采样率（Hz，如 44100） */
    public int sampleRate;

    /** 文件名（UTF-8） */
    public String name;

    /** 标题（UTF-8） */
    public String title;

    /** 艺术家（UTF-8） */
    public String artist;

    /** 专辑（UTF-8） */
    public String album;

    /** 所在内存单元（0=内置闪存, 1=SD 卡） */
    public byte memUnit;

    /**
     * 创建一个空的 {@code Song}
     */
    public Song() {
        this.fileNo = 0;
        this.size = 0;
        this.time = 0;
        this.bitRate = 0;
        this.sampleRate = 0;
        this.name = "";
        this.title = "";
        this.artist = "";
        this.album = "";
        this.memUnit = 0;
    }

    @Override
    public String toString() {
        return "Song{"
                + "fileNo=" + fileNo
                + ", size=" + size
                + ", time=" + time
                + ", bitRate=" + bitRate
                + ", sampleRate=" + sampleRate
                + ", name='" + name + "'"
                + ", title='" + title + "'"
                + ", artist='" + artist + "'"
                + ", album='" + album + "'"
                + ", memUnit=" + memUnit
                + "}";
    }
}

File: platforms/android-native/app/src/main/java/c/cyrio/android/model/Song.kt
package c.cyrio.android.model

import org.json.JSONObject

/**
 * 歌曲数据模型
 *
 * 对应 Rust 侧 SongJson：
 * {"fileNo","size","time","bitRate","sampleRate","name","title","artist","album","memUnit"}
 */
data class Song(
    val fileNo: Int,
    val size: Int,
    val time: Int,
    val bitRate: Int,
    val sampleRate: Int,
    val name: String,
    val title: String,
    val artist: String,
    val album: String,
    val memUnit: Int
) {
    /** 文件大小格式化（KB/MB） */
    val sizeText: String
        get() = formatSize(size)

    /** 时长格式化 mm:ss */
    val timeText: String
        get() {
            val m = time / 60
            val s = time % 60
            return "%d:%02d".format(m, s)
        }

    /** 比特率格式化 */
    val bitRateText: String
        get() = "${bitRate / 1000}kbps"

    /** 存储位置文本 */
    val memUnitText: String
        get() = if (memUnit == 0) "内置" else "SD卡"

    companion object {
        /** 从 JSON 数组字符串解析歌曲列表 */
        fun parseList(json: String): List<Song> {
            if (json.isBlank() || json.trim() == "[]") return emptyList()
            val arr = org.json.JSONArray(json)
            val result = ArrayList<Song>(arr.length())
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                result.add(parse(o))
            }
            return result
        }

        /** 从 JSONObject 解析单首歌曲 */
        fun parse(o: JSONObject): Song {
            return Song(
                fileNo = o.optInt("fileNo"),
                size = o.optInt("size"),
                time = o.optInt("time"),
                bitRate = o.optInt("bitRate"),
                sampleRate = o.optInt("sampleRate"),
                name = o.optString("name"),
                title = o.optString("title"),
                artist = o.optString("artist"),
                album = o.optString("album"),
                memUnit = o.optInt("memUnit")
            )
        }

        /** 格式化文件大小 */
        fun formatSize(bytes: Int): String {
            if (bytes < 1024) return "${bytes}B"
            if (bytes < 1024 * 1024) return "%.1fKB".format(bytes / 1024.0)
            return "%.1fMB".format(bytes / (1024.0 * 1024.0))
        }
    }
}
